### PR TITLE
Introducing an explicit `Declare Scope` command

### LINKED ISSUE #7120: Interpretation scopes should be declared
Currently, a scope is automatically created with the first of `Notation "a" := b : scope`, `Delimit scope with key`, or `Bind Scope scope with T`.
(are there other ways to declare a scope?)
In contrast, `Open Scope scope` does not create scope.

@RalfJung commented (https://github.com/coq/coq/pull/6247#issuecomment-377202990) that the current implicit binding is awkward, since typos will silently create a new scope. I agree.
Sometimes you also want to declare a sequence of notations in a new scope:
```
Notation "+" := ( 1 ) : AA.
Open Scope AA.
Notation "*" := ( + ) : AA.
```
Here, `Open Scope AA` has to come *second* in the list of commands: it makes more sense to have it at the top, but `AA` hasn't been declared yet, so you get an error.

I propose adding a new command `Declare Scope scope` which is equivalent to the commands `Delimit Scope scope with placeholder. Undelimit scope.`,
and adding a warning whenever a new scope is created other than by `Declare Scope`, with the intention to turn the warning into an error after a suitable deprecation period.

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** feature / enhancement / partial redesign / wish

In some sense: "closes #7120".

This PR adds a command `Declare Scope id` to register a scope and displays a warning the first time a non-explicitly registered scope is referred to.

This answers a [suggestion](https://github.com/coq/coq/pull/6247#issuecomment-377202990) initially made in the discusion for #6247 and turned into a proper issue #7120 by @jashug.

As far as I could judge (hoping I did not miss a case), behavior of existing scripts should be unchanged up to the new output of a warning. As for the new command, it makes a scope visible at require time as it was before for scopes implicitly declared by either `Delimit`, `Undelimit`, `Bind` or `Notation`.

Part of the remaining issues are discussed at #7120. Wrt to @jashug's [comment](), the current PR deactivates warning at require time, except for primitive notation (more difficult).

To do:
- [X] Corresponding documentation was added / updated.
- [x] Entry added in CHANGES.

Possibly to do:
- [ ] unify the different definitions of a scope `pair_scope` (e.g. by moving the notations for `fst` and `snd` in `Datatypes.v`, also giving notations for `projT1` and `projT2` at the same time)?
- [ ] define a common scope for `CMorphisms` and `Morphisms`, as well as `CEquivalences` and `Equivalences`?
- [ ] unify names of scopes (with e.g. `_scope` convention) in standard library

### CODE REVIEW

**Comment by gares** [comment]
While you are at it, why don't you let one declare in one go both the scope and its delimiter?
```
Declare Scope nat_scope as N.
```
If you look at your patch to ssrmatching.v you see why I ask for this.

BTW, why do we need both concepts?

**Comment by herbelin** [comment]
Sure, this is a recurrent question. Do you have a concrete proposal?

**Comment by Zimmi48** [comment]
Isn't `Declare Scope nat_scope as N.` concrete enough?

**Comment by herbelin** [comment]
> Declare Scope nat_scope as N.

Ah, sorry, I misunderstood. I thought it was about avoiding the redundancy between the name `nat_scope` and the delimiter, a redundancy which I find a bit tedious.

So, about a `Declare Scope nat_scope as N` proposal, the question is whether we systematically want a delimiter. If it is clear that we don't want a scope w/o a delimiter, we can combine the two commands into one. And if it is clear that we have a bijection between delimiters and scope names, there is then the question of redundancy between the scope name and the delimiter.

So, in short, the questions seem to me to be:
- Do we want to enforce (bijectively) a delimiter? If yes, we could for instance adopt a convention that the name of a scope is the delimiter name prefixed by `%`. So, we would have for instance: `Declare Scope %nat.`, `Delimit Scope %nat with nat.` and `Notation "x + y" := (Nat.add x y) : %nat.`.
- Do we even want, maybe, to have a scope canonically associated to types. In which case, we could have that every time constant `foo` is defined (*), there is an associated scope and delimiter `%foo` bound to it. In this case, `%nat` would exist by default as soon as `nat` is defined and we would need a `Declare Scope bar.` only for scopes not already canonically bound to a constant. It is however unclear how to build something roughly compatible with the existing by following something along this second line. For instance, since any `%foo` would exist by default, that would make the space of available delimiter names much limited (e.g. `%pair` could not be used as the name of a new scope for projections, since it would already be by default bound  to the constant `pair`... or to not associate a scope to `pair` by default on the reason that it does not build a type, but then this starts to be a complicated rule...).

(*) or at least a constant liable to build a typetype `N`.

**Comment by gares** [comment]
I'm sorry I mixed two topics.

1. in the files I'm asked to review I see a `Declare sope` line immediately followed by a `Delimit scope` one, hence my proposal `Declare Scope XXX as YYY`. The `as ...` part could be optional, in case one wants no delimiter.
2. I never fully understood why we both have a scope name and a scope delimiter. And the obvious proposal is to just have one that is used both for notations `... : nat` and inside terms prefixed by `%`. The answer "you may want to have no delimiter displayed but still put the notation in a scope" is OK for me.

**Review by gares — APPROVED** [review]
I'm ok with the changes in `plugins/ssr/`.
I think having a single command to declare a scope and its delimiter would make our users happy(er) to fix the warning.

**Comment by ejgallego** [comment]
> define a common scope for CMorphisms and Morphisms, as well as CEquivalences and Equivalences?

So what's the semantics of declaring a scope twice?

**Comment by ejgallego** [comment]
> I think having a single command to declare a scope and its delimiter would make our users happy(er) to fix the warning.

Good point, is it mandatory for a scope to have a delimiter?

**Comment by ejgallego** [comment]
If you go with the route of making scope declaration mandatory, then I'd suggest you re-utilize the vernac for bind scope.

BTW: who should be assigned to this PR.

**Review by jashug — APPROVED** [review]
Looks good to me.

**Comment by herbelin** [comment]
> So what's the semantics of declaring a scope twice?

Currently, in this PR, the second time is a no-op.

> Good point, is it mandatory for a scope to have a delimiter?

A command `Undelimit` has been introduced at some time (#83), so I guess there is some interest for it (though maybe, the intent is not per se to undelimit but to deactivate the use of a scope).

**Comment by ejgallego** [comment]
> Currently, in this PR, the second time is a no-op.

I think that'd be fine. Still not clear I am gonna be the responsible person for merging this. If so, it looks OK to me other than leaving to more expert notations users to decide about whether an scope should always have a delimiter or not, as I feel that would allow to simplify code and vernaculars a bit.

**Comment by herbelin** [comment]
Maybe some more time needed before taking decisions about the following open points:
- Is it worth to introduce a new way to define both a scope and a delimiter, using the notation %foo for the name of the scope with delimiter foo? Or is the burden of the change not worth the advantage of removing the redundancy between scope name and delimiter?
- In a context where delimiter and scope name are merged, is it worth to rethink at `Undelimit`as `Deactivate`? Or would it be again too much change for minor added value?
- If we don't merge scope name and delimiter, does it remain worth to have a command to define both at the same time, since this command will have anayway to exist also in a variants where only one of them is defined at a time?

**Comment by Zimmi48** [comment]
Maybe a poll on Coq-Club would be useful to get users' opinion. I'm personally not a big notation user, but for teaching I would prefer the simplicity of having the scope and the delimiter merged to a single thing.
If merging was decided, we would need a transition strategy as well.

**Comment by gares** [comment]
The transition could be:
- today (Coq 8.9): you give a warning if there is no explicit Declare (be it a separate line or glued with the delimiter declaration). It is a warning the user can fix.
- tomorrow (Coq 8.10): you turn it into an error
Am I missing a subtlety?

**Comment by Zimmi48** [comment]
I was taking about a transition strategy in the case it was decided to merge scope names and scope delimiters. But maybe I am the one who misunderstood what was being discussed...

**Comment by ejgallego** [comment]
Well it looks lie noboby commented on the "who's driving this" so taking over for now. I will make summary with action plan soon, while @herbelin writes the docs, things about the possible choices.

But for the moment let me suggest that @herbelin submits the first part of this PR, splitting the primitive notation plugins as a separate PR, so we can merge it quick.

We already had this discussion before, and IMHO adding unrelated changes to the PRs complicates things. Even if I understand Hugo very well as the same often happens to me, and having to submit and manage all these depending PRs is a bit tiring.

**Comment by herbelin** [comment]
I made the split of primitive parsers/printers a separate PR. So, this PR is now dependent on it, namely #7410.

I'm otherwise missing time, so unless someone is willing to take over, I believe I cannot propose better than the current state of the PR (i.e. providing a `Declare Scope` command, but not merging it with delimiter name nor delimiter declaration; also giving a warning if a scope is used without being defined).

**Comment by ejgallego** [comment]
> I made the split of primitive parsers/printers a separate PR. So, this PR is now dependent on it, namely #7410.

Cool, can you at least rebase it on top, I guess #7410 will be merged without problems.

**Comment by herbelin** [comment]
> can you at least rebase it on top

Ja. Have forgotten to push actually.

**Comment by ejgallego** [comment]
Danke schön!

**Comment by ejgallego** [comment]
I need to look at the patch with care myself, in principle I don't see any problem to merge this.

Meanwhile I'm gonna leave a few days so people can comment.

**Comment by ejgallego** [comment]
@herbelin it seems we are missing the addition to the reference manual. Do you think you could push an additional commit adding an entry there? Thanks.

**Comment by ejgallego** [comment]
It also needs a (trivial) rebase.

File: interp/notation.ml
Comment (gares) [inline interp/notation.ml]: This looks like copy paste, I think `declare_scope` should take a flag to eventually raise the warning.

File: plugins/setoid_ring/Ncring_initial.v
Comment (JasonGross) [inline plugins/setoid_ring/Ncring_initial.v]: I think this should be ` Local Declare Scope`

**Review by JasonGross — DISMISSED** [review]
Does this conflict with numeral notations, or is that left over from a missing rebase?

File: plugins/setoid_ring/Ncring_initial.v
Comment (herbelin) [reply plugins/setoid_ring/Ncring_initial.v]: OK, fixed.

File: interp/notation.ml
Comment (herbelin) [reply interp/notation.ml]: I tried to use a flag but I'm not really satisfied by what I got. Tell me which version you prefer.

File: plugins/setoid_ring/Ncring_initial.v
Comment (herbelin) [reply plugins/setoid_ring/Ncring_initial.v]: I answered too quickly. We should first give a semantics to `Local Declare Scope` in sections.

For `Notation` and `Open Scope`/`Close Scope`, `Local` is a no-op in sections.

For `Notation`, `Global` is forbidden in sections.

For `Open Scope`/`Close Scope`, `Global` tells to survive sections.

I see two possible choices:
- do like `Notation`, i.e. forbid `Global` in sections, and have `Local` a no-op
- do like `Open Scope`, i.e. allow `Global` in sections to tell discharge and have `Local` a no-op

Note that for `Delimit Scope`, `Undelimit Scope` and `Bind Scope`, `Local` and `Global` are currently forbidden in sections but we could seize this opportunity to support them the same way as we do for `Declare Scope`.

**Comment by herbelin** [comment]
> What are the scoping rules of Declare Scope?

You mean wrt sections and modules? It uses `enforce_module_locality`, so it defaults to export in modules (unless `Local` is used) and in discard in sections (with `Global` not accepted). So, nothing changed relatively to the discussion [above](https://github.com/coq/coq/pull/7135#discussion_r206079380).

@jasongross: I finally removed the useless `Local` in `Ncring_initial.v`.

**Comment by ejgallego** [comment]
Build seems to be failing.

**Comment by herbelin** [comment]
I missed a compilation warning. Retrying.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: Are these built-in scopes or scopes declared in the prelude?

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: "in above": remove "in".

File: interp/notation.ml
Comment (Zimmi48) [inline interp/notation.ml]: I suggest using `~category:"deprecated"` so that people can easily make all deprecated warnings fatal.

**Review by Zimmi48 — APPROVED** [review]
Approval of the doc part. However, note my remark on the deprecation warning. I also suggest that you document this new warning in the manual (use the `.. warn::` directive).

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: All names are declared in the prelude, but some have a specific built-in action corresponding to the class they are bound to (e.g. `CL_SORT` <-> `type_scope`, and `CL_FUN` <-> `function_scope`).

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (herbelin) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: Well spotted. Thanks

File: interp/notation.ml
Comment (herbelin) [reply interp/notation.ml]: OK, doing it, but maybe time to move to multi-category warnings.

**Comment by ejgallego** [comment]
Ok, so this finally seems ready to merge. People had plenty of time to comment so after a last review from me I will merge tomorrow.

File: CHANGES
Comment (ejgallego) [inline CHANGES]: I am much afraid that this entry is placed in the wrong place, please fix @herbelin

**Review by ejgallego — DISMISSED** [review]
Cool stuff but unfortunately the entry in `CHANGES` looks wrong.

File: CHANGES
Comment (herbelin) [reply CHANGES]: Right! Moved to post-8.8.2 (BTW, who is supposed to eventually restructurate the `CHANGES`, e.g. to glue the two post-8.8.2 sections `Notations`, etc., shall it be @silene?)

**Review by ejgallego — APPROVED** [review]

**Comment by Zimmi48** [comment]
>Moved to post-8.8.2 (BTW, who is supposed to eventually restructurate the CHANGES, e.g. to glue the two post-8.8.2 sections Notations, etc., shall it be @silene?)

Yes, until we move to an automated generation of the changelog, the RM will have some work to do to fix CHANGES before a release. Something that I forgot from the release check-list apparently. People are still putting features in the 8.9+beta1 milestone and the 8.9+beta1 part of CHANGES even if the branching has happened so there will be untangling work to do on this as well.

**Comment by ejgallego** [comment]
> People are still putting features in the 8.9+beta1 milestone and the 8.9+beta1 part of CHANGES even if the branching has happened so there will be untangling work to do on this as well.

It seems to me that's not the fault of the people doing that, we are getting scarce, if any information on what the beta process is going to be.

**Comment by Zimmi48** [comment]
Agreed.

**Pull request merged**
